src/sfizz/Synth.cpp

```diff
@@ -577,6 +578,94 @@ int sfz::Synth::getNumCurves() const noexcept
     return numCurves;
 }
 
+std::string sfz::Synth::exportMidnam(absl::string_view model) const
+{
+    pugi::xml_document doc;
+
+    // TODO: perhaps define this somewhere, or use some metadata from the SFZ
+    const char *manufacturer = "Paul Ferrand";
+    const char *defaultModel = "Sfizz";
+
+    if (model.empty())
+        model = defaultModel;
+
+    doc.append_child(pugi::node_doctype).set_value(
+        "MIDINameDocument PUBLIC"
+        " \"-//MIDI Manufacturers Association//DTD MIDINameDocument 1.0//EN\""
+        " \"http://www.midi.org/dtds/MIDINameDocument10.dtd\"");
+
+    pugi::xml_node root = doc.append_child("MIDINameDocument");
+
+    root.append_child(pugi::node_comment)
+        .set_value("Generated by Sfizz for the current instrument");
+
+    root.append_child("Author");
+
+    pugi::xml_node device = root.append_child("MasterDeviceNames");
+    device.append_child("Manufacturer")
+        .append_child(pugi::node_pcdata).set_value(manufacturer);
+    device.append_child("Model")
+        .append_child(pugi::node_pcdata).set_value(std::string(model).c_str());
+
+    {
+        pugi::xml_node devmode = device.append_child("CustomDeviceMode");
+        devmode.append_attribute("Name").set_value("Default");
+
+        pugi::xml_node nsas = devmode.append_child("ChannelNameSetAssignments");
+        for (unsigned c = 0; c < 16; ++c) {
+            pugi::xml_node nsa = nsas.append_child("ChannelNameSetAssign");
+            nsa.append_attribute("Channel").set_value(std::to_string(c + 1).c_str());
+            nsa.append_attribute("NameSet").set_value("Play");
+        }
+    }
+
+    {
+        pugi::xml_node chns = device.append_child("ChannelNameSet");
+        chns.append_attribute("Name").set_value("Play");
+
+        pugi::xml_node acs = chns.append_child("AvailableForChannels");
+        for (unsigned c = 0; c < 16; ++c) {
+            pugi::xml_node ac = acs.append_child("AvailableChannel");
+            ac.append_attribute("Channel").set_value(std::to_string(c + 1).c_str());
+            ac.append_attribute("Available").set_value("true");
+        }
+
+        chns.append_child("UsesControlNameList")
+            .append_attribute("Name").set_value("Controls");
+    }
+
+    {
+        pugi::xml_node cns = device.append_child("ControlNameList");
+        cns.append_attribute("Name").set_value("Controls");
+        for (const CCNamePair& pair : ccNames) {
+            pugi::xml_node cn = cns.append_child("Control");
+            cn.append_attribute("Type").set_value("7bit");
+            cn.append_attribute("Number").set_value(std::to_string(pair.first).c_str());
+            cn.append_attribute("Name").set_value(pair.second.c_str());
+        }
+    }
+
+    ///
+    struct string_writer : pugi::xml_writer {
+        std::string result;
+
+        string_writer()
+        {
+            result.reserve(8192);
```

Comment:
Hosts will call this `exportMidnam()` method outside of the audio thread I believe?

Reply:
There is no way this is designed to occur in realtime context, given this function is required to allocate memory for the text.